### ISSUE TITLE
Exemplars are not recorded for metric `http_server_requests*`

### ISSUE THREAD

**Issue by ahus1** [quarkusio/quarkus#43845]
### Describe the bug

I'm running Keycloak with Metrics and OpenTelemetry tracing enabled. 

When running HTTP 1.x requests, I find that the metrics are recorded for `http_server_requests_*`, but no exemplars are recorded. Other metrics have exemplars. 

NOTE: I retrieved metrics from my application using 

```
curl http://localhost:9000/metrics
```

as when running with Chrome, it never returned the OpenMetrics format with the Exemplars, but always the old text/plain Prometheus format (possibly due to content negotiation in the headers). 

### Expected behavior

I expect exemplars to be provided for http server metrics.

```
http_server_requests_seconds_count{method="GET",outcome="SUCCESS",status="204",uri="/realms/{realm}/protocol/{protocol}/login-status-iframe.html/init"} 2.0 # {span_id="cc258ebf850c7ec8",trace_id="39750627ed1138e074b6256816bfa845"} 1.0 1728839669.614
```

### Actual behavior

They are missing

```
http_server_requests_seconds_countmethodGEToutcomeREDIRECTIONstatus302uri/admin 3.0
http_server_requests_seconds_summethodGEToutcomeREDIRECTIONstatus302uri/admin 0.0324834
http_server_requests_seconds_countmethodGEToutcomeSUCCESSstatus204uri/realms/{realm}/protocol/{protocol}/login-status-iframe.html/init 7.0
```

### How to Reproduce?

* Start Keycloak 26.0.0 and enable Metrics and OpenTelementry
* Login to issue some HTTP requests
* Retrieve metrics from the metrics endpoint using CURL

### Output of `uname -a` or `ver`

Windows 11

### Output of `java -version`

JDK 17

### Quarkus version or git rev

3.15.1

### Additional information

I found that `OpentelemetryExemplarSamplerProvider` expects the current trace context to be present in the Vert.x context:

https://github.com/quarkusio/quarkus/blob/main/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/OpentelemetryExemplarSamplerProvider.java#L25-L40

Unfortunately Vert.x doesn't set the context:

https://github.com/eclipse-vertx/vert.x/blob/1e2fdf820be9628af2576c0f32828802111855b1/vertx-core/src/main/java/io/vertx/core/http/impl/Http1xServerConnection.java#L265-L270

I assume this can be fixed by changing Vert.x to set the context. I'll open an issue there to suggest a change. I'll post a link to the issue here. 

Happy for other pointers and ideas from the Quarkus team. Thanks!

**Comment by tsegismont** [comment]
`io.vertx.core.http.impl.Http1xServerConnection#reportResponseComplete` invocation does not happen on a Vert.x Context at all: it's a Netty event-loop task that executes this task (for good reasons).

In `reportResponseComplete`, the tracer is invoked:

```java
    if (tracer != null && trace != null) {
      tracer.sendResponse(request.context, request.response(), trace, null, HttpUtils.SERVER_RESPONSE_TAG_EXTRACTOR);
    }
```

The context bound to the request is provided to the tracing implementation (first param).

I'm not knowledgeable with the tracing implementation in Quarkus, but I would say that instead of expecting to find tracing data on the current context, it should look it up in the context provided as a param.

**Comment by ahus1** [comment]
Minimal reproducer project available here: https://github.com/ahus1/quarkus-otel-missing-examplar-example

**Comment by ahus1** [comment]
@tsegismont - the exemplar is recorded when the metric is recorded, not when the trace is recorded. 

See https://github.com/eclipse-vertx/vert.x/pull/5356 for a PR I prepared for upstream.

**Comment by tsegismont** [comment]
Also, you can capture the context bound to a request with the metrics SPI.

When the `requestBegin` method is invoked:

```java
    if (metrics != null) {
      metric = metrics.requestBegin(conn.metric(), this);
    }
```

The second parameter is the HTTP request and can be cast to `HttpServerRequestInternal`. This type exposes the context bound to the request (and then you can store it in the request metric object returned by `requestBegin`.

**Comment by tsegismont** [comment]
In general, it shouldn't be assumed that metrics SPI methods will be invoked on a Vert.x context. They can be invoked directly by the EL thread managing the connection.

**Comment by tsegismont** [comment]
> [@tsegismont](https://github.com/tsegismont) - the exemplar is recorded when the metric is recorded, not when the trace is recorded.

Yeah, sorry, I missed that initially and then completed my answer while you were typing :-)

**Comment by ahus1** [comment]
FYI @vmuzikar, @mabartos: I found that exemplars are not working for HTTP metrics when tracing is enabled in Keycloak 26.

Having a an exemplar would allow you to link a metrics counter of HTTP 500s or a metrics bucket of slow requests to a recorded trace which would show you the root cause of the error, or the reason why it is slow.

**Comment by ahus1** [comment]
@tsegismont - from [your comment on the Vert.x project](https://github.com/eclipse-vertx/vert.x/issues/5355#issuecomment-2418986708) I understand that you and the Quarkus metrics experts are looking into a Quarkus option on how to fix this. I'll pause here to see suggestions.

**Comment by brunobat** [comment]
@ahus1 I meet with @tsegismont and he mentions that the proposed solution for vert.x will have a big performance.
I will explore retrieval and set of the OTel context to set that information on the Quarkus side.

**Comment by tsegismont** [comment]
For the record, the idea is to retrieve the OTel context in `io.quarkus.micrometer.runtime.binder.vertx.VertxHttpServerMetrics#responseEnd` and somehow to _attach_ it to the thread invoking `sample.stop`.

```java
  Context otelContext = QuarkusContextStorage.getContext(requestMetric.request().context());
  try {
      // attach
      sample.stop(requestsTimer.withTags());
  } finally {
      // detach
  }
```

I'm not familiar with OTel API so not sure how to do that exactly

**Comment by brunobat** [comment]
The code seems to work:

> http_server_requests_seconds_count{method="GET",outcome="SUCCESS",status="200",uri="root"} 2.0 # {span_id="068690c2a723d8ca",trace_id="d15ed60e6c707dcf653195c58c913347"} 1.0 1729265792.611

But I need to create tests from scratch.
There are no tests for exemplars right now.
Will continue next week.
